hnn_core/mpi_child.py

```diff
@@ -18,79 +20,123 @@ def _read_all_bytes(stream_in, chunk_size=4096):
+            sys.stdin.close()
+
+            params = pickle.loads(codecs.decode(input_bytes, "base64"))
+        else:
+            params = None
+
+        params = self.comm.bcast(params, root=0)
+        return params
+
+    def _write_data_stderr(self, sim_data):
+        """write base64 encoded data to stderr"""
+
+        # only have rank 0 write to stdout/stderr
+        if self.rank > 0:
+            return
 
         # pickle the data and encode as base64 before sending to stderr
         pickled_str = pickle.dumps(sim_data)
-        pickled_bytes = codecs.encode(pickled_str,
-                                      'base64')
+        pickled_bytes = codecs.encode(pickled_str, 'base64')
+
+        # base64 encoding requires data padded to a multiple of 4
+        padding = len(pickled_bytes) % 4
+        pickled_bytes += b"===" * padding
         sys.stderr.write(pickled_bytes.decode())
         sys.stderr.flush()
 
-    MPI.Finalize()
-    return 0
+        # the parent process is waiting for "end_of_sim:[#bytes]" with the
+        # length of data
+        sys.stdout.write('end_of_data:%d' % len(pickled_bytes))
+        sys.stdout.flush()  # flush to ensure signal is not buffered
+
+    def run(self, params):
+        """Run MPI simulation(s) and write results to stderr"""
+
+        from hnn_core import Network
+        from hnn_core.parallel_backends import _clone_and_simulate
+
+        prng_seedcore_initial = params['prng_*']
+
+        net = Network(params)
+        sim_data = []
+        for trial_idx in range(params['N_trials']):
+            single_sim_data = _clone_and_simulate(net, trial_idx,
+                                                  prng_seedcore_initial)
+
+            # go ahead and append trial data for each rank, though
+            # only rank 0 has data that should be send back to MPIBackend
+            sim_data.append(single_sim_data)
+
+        # flush output buffers from all ranks (any errors or status mesages)
+        sys.stdout.flush()
+        sys.stderr.flush()
+
+        if self.rank == 0:
+            # the parent process is waiting for "end_of_sim" to signal that
+            # the following stderr will only contain sim_data
+            sys.stdout.write('end_of_sim')
+            sys.stdout.flush()  # flush to ensure signal is not buffered
+
+        return sim_data
 
 
 if __name__ == '__main__':
+    """This file is called on command-line from nrniv"""
+
+    import traceback
+    rc = 0
+
     try:
-        rc = run_mpi_simulation()
-    except Exception as e:
+        with MPISimulation() as mpi_sim:
+            params = mpi_sim._read_params()
+            sim_data = mpi_sim.run(params)
+            mpi_sim._write_data_stderr(sim_data)
+    except Exception:
         # This can be useful to indicate the problem to the
         # caller (in parallel_backends.py)
-        print("Exception: %s" % e)
+        traceback.print_exc(file=sys.stdout)
```

Comment:
does this mean we'll now get a full traceback? Why do you catch the exception then? to send it to stdout?

Reply:
Yeah, stdout is sure to get printed to the screen by MPIBackend. Stderr might be interpreted as data (which would fail, but the traceback wouldn't be shown).

Reply:
I see. I am not super clear as to what goes to stdout and what goes to stderr etc. Should we perhaps document it somewhere for future contributors? Just a few lines, nothing too long. The code is there of course :)